src/cmd/ksh93/sh/init.c

```diff
@@ -907,25 +907,46 @@ static_fn char *get_version(Namval_t *np, Namfun_t *fp) { return (nv_getv(np, fp
 // This is invoked when var `.sh.version` is used in a numeric context such as
 // `$(( .sh.version ))`.
 //
-// TODO: Figure out what we want this to do with a semantic version string and
-// a version string that may include information about the git commit used for
-// the build. The original version converted the YYYY-MM-DD date stamp into the
-// number represented by YYYYMMDD. But note it would return a nonsensical value
-// if a date stamp in the correct form did not start exactly 10 characters
-// before the end of the string referred to by`e_version`.
 static Sfdouble_t nget_version(Namval_t *np, Namfun_t *fp) {
     UNUSED(np);
-    const char *cp = SH_RELEASE;
-    int c;
+    char *cp = strdup(SH_RELEASE);
     Sflong_t t = 0;
+    char *dash;
+    char *dot;
+    char *major_str, *minor_str, *patch_str;
+    int major, minor, patch;
     UNUSED(fp);
 
-    while ((c = *cp++)) {
-        if (c >= '0' && c <= '9') {
-            t *= 10;
-            t += c - '0';
-        }
-    }
+    // Version string in development version could be set like 2017.0.0-devel-1509-g95d59865
+    // If a '-' exists in version string, set it as end of string i.e. version string becomes
+    // 2017.0.0
+    dash = strchr(cp, '-');
+    if (dash) *dash = 0;
+
+    // Major version number starts at beginning of string
+    major_str = cp;
+
+    // Find the first '.' and set it to NULL, so major version string is set to 2017
+    dot = strchr(cp, '.');
```

Comment:
I just noticed this sequence of statements isn't safe if the `SH_RELEASE` string doesn't contain a period in which case `dot` is NULL. This needs to be hardened against that scenario. We should probably just return zero because I can't imagine any other value that would not mislead or cause other problems.

Reply:
And I just tested this change and it does, as expected, generate a SIGSEGV on the next statement using my repo which does not have a semantic version tag.

Reply:
I modified code to use old style of conversion if `.` is not present in version number.